homeassistant/components/sense/__init__.py

```diff
@@ -39,34 +43,90 @@
 )
 
 
-async def async_setup(hass, config):
-    """Set up the Sense sensor."""
+async def async_setup(hass: HomeAssistant, config: dict):
+    """Set up the Sense component."""
+    hass.data.setdefault(DOMAIN, {})
+    conf = config.get(DOMAIN)
+    if not conf:
+        return True
+
+    hass.async_create_task(
+        hass.config_entries.flow.async_init(
+            DOMAIN,
+            context={"source": SOURCE_IMPORT},
+            data={
+                CONF_EMAIL: conf.get(CONF_EMAIL),
+                CONF_PASSWORD: conf.get(CONF_PASSWORD),
+                CONF_TIMEOUT: conf.get(CONF_TIMEOUT),
+            },
+        )
+    )
+    return True
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry):
+    """Set up Sense from a config entry."""
+
+    hass.data.setdefault(DOMAIN, {})
 
-    username = config[DOMAIN][CONF_EMAIL]
-    password = config[DOMAIN][CONF_PASSWORD]
+    entry_data = entry.data
+    email = entry_data[CONF_EMAIL]
+    password = entry_data[CONF_PASSWORD]
+    timeout = entry_data[CONF_TIMEOUT]
 
-    timeout = config[DOMAIN][CONF_TIMEOUT]
     try:
-        hass.data[SENSE_DATA] = ASyncSenseable(api_timeout=timeout, wss_timeout=timeout)
-        hass.data[SENSE_DATA].rate_limit = ACTIVE_UPDATE_RATE
-        await hass.data[SENSE_DATA].authenticate(username, password)
+        gateway = ASyncSenseable(api_timeout=timeout, wss_timeout=timeout)
+        gateway.rate_limit = ACTIVE_UPDATE_RATE
+        await gateway.authenticate(email, password)
     except SenseAuthenticationException:
         _LOGGER.error("Could not authenticate with sense server")
         return False
-    hass.async_create_task(async_load_platform(hass, "sensor", DOMAIN, {}, config))
-    hass.async_create_task(
-        async_load_platform(hass, "binary_sensor", DOMAIN, {}, config)
-    )
+
+    hass.data[DOMAIN][entry.entry_id] = {}
+    hass.data[DOMAIN][entry.entry_id][SENSE_DATA] = gateway
```

Comment:
This is obviously fine, you can also combine the 2 statements:

```suggestion
    hass.data[DOMAIN][entry.entry_id] = {SENSE_DATA: gateway}
```

Reply:
Adjusted